server/data_common/utils.py

```diff
@@ -0,0 +1,90 @@
+from functools import wraps
+
+from flask import json
+from numpy import float32, integer
+
+from server.common.errors import DriverError
+from enum import Enum
+
+
+class Float32JSONEncoder(json.JSONEncoder):
+    def __init__(self, *args, **kwargs):
+        """
+        NaN/Infinities are illegal in standard JSON.  Python extends JSON with
+        non-standard symbols that most JavaScript JSON parsers do not understand.
+        The `allow_nan` parameter will force Python simplejson to throw an ValueError
+        if it runs into non-finite floating point values which are unsupported by
+        standard JSON.
+        """
+        kwargs["allow_nan"] = False
+        super().__init__(*args, **kwargs)
+
+    def default(self, obj):
+        if isinstance(obj, float32):
+            return float(obj)
+        elif isinstance(obj, integer):
+            return int(obj)
+        return json.JSONEncoder.default(self, obj)
+
+
+def custom_format_warning(msg, *args, **kwargs):
+    return f"[cellxgene] Warning: {msg} \n"
+
+
+def jsonify_numpy(data):
+    return json.dumps(data, cls=Float32JSONEncoder, allow_nan=False)
+
+
+def requires_data(func):
```

Comment:
this function seems specific to the scanpy_engine, as it references a field (`data`) which is unique to that driver sub-class

Reply:
I removed this.  It was part of a feature that allowed a ScanpyEngine to be created but not initialized.  That feature was never used (except of a test written specifically to test that it worked).  So I also removed that delayed initialization feature.